Adding a tutorial for graph regularization with images

sayakpaul opened the pull request:
@arjung I reckoned that adding the example under `g3docs/tutorials` might be more appropriate because it would be the first one that shows how to build a synthetic graph for an image dataset and build a model. 

Let me know your thoughts. A parallel Colab Notebook is available [here](https://colab.research.google.com/gist/sayakpaul/1a8e5202fe84bbc2b0e6bb38500109e8/graph_regularization_images.ipynb) for the results. For reference, the corresponding issue thread is available [here](https://github.com/tensorflow/neural-structured-learning/issues/87).

arjung commented on g3doc/tutorials/graph_keras_cnn_flowers.ipynb:
```diff
@@ -0,0 +1,1198 @@
+        "This notebook is a sister notebook to this example - [Graph regularization for sentiment \n",
+        "classification using synthesized graphs](https://www.tensorflow.org/neural_structured_learning/tutorials/graph_keras_lstm_imdb). In here, we'll build a flower classification model that can\n",
+        "categorize images of flowers into discrete classes such as sunflowers, roses,\n",
+        "tulips, etc. \n",
```
Reword this sentence as "This notebook is the counterpart of [Graph regularization for sentiment classification using synthesized graphs](https://www.tensorflow.org/neural_structured_learning/tutorials/graph_keras_lstm_imdb) for image classification. In this notebook, we will build a flower classification model to categorize images of flowers into discrete classes such as sunflowers, roses, tulips, and so on."

arjung commented on g3doc/tutorials/graph_keras_cnn_flowers.ipynb:
```diff
@@ -0,0 +1,1198 @@
+        "id": "cShQNXGNjIOs"
+      },
+      "source": [
+        "Based on the performances of the two models (with and without graph regularization) we observe that the model performs better without graph regularization. But an ablation study (on the reduced projection dimensionality) reveals a different trend\n",
```
This is not a good message for prospective users. Can we we set the projection dimensionality to a value that gives maximum improvement for graph regularization?

arjung commented on g3doc/tutorials/graph_keras_cnn_flowers.ipynb:
```diff
@@ -0,0 +1,1198 @@
+        "| 384 \t| 58% \t| 60.18% \t|\n",
+        "| 512 \t| 57.45% \t| 57.64% \t|\n",
+        "\n",
+        "The table above reports the validation top-1 accuracies with and without graph regularization under different reduced embedding dimensionalities. Notice that after 256-d the performance trade-off reverses. Maybe a bit more experimentation with hyperparameters like `similarity_threshold` and `graph_regularization_multiplier` is needed to fully study this behavior. "
```
Do you plan to do more work to understand this? If not, I don't know if we need this statement that 'more experimentation is needed....'

arjung commented on g3doc/tutorials/graph_keras_cnn_flowers.ipynb:
```diff
@@ -0,0 +1,1198 @@
+      "source": [
+        "## Visualization\n",
+        "\n",
+        "In this final section, we visualize the neighbor images as computed by Neural Structured Learning during building the graph. "
```
Nice!

arjung reviewed (commented):
Thanks for the PR, @sayakpaul! Looks generally very good! Most of my comments are on understandability and documentation, and should be straightforward to address. If possible, try *not* to sync your local workspace to HEAD prior to sending the next version on this PR.

arjung reviewed (changes requested):
One other thing I forgot to add earlier. Please move this notebook to https://github.com/tensorflow/neural-structured-learning/tree/master/neural_structured_learning/examples/notebooks like I mentioned initially. We also discussed this last week with the team again and decided that that would be a better destination for this tutorial for now.

sayakpaul commented:
@arjung thanks for your feedback but I am not comfortable with moving this to the location you suggested. The primary reason is visibility. 

So, if the team decides to consider that later at some point in time we can keep this PR open otherwise, closing the PR seems to be the only option.

arjung commented:
I thought this was something we agreed upon prior to starting work on this tutorial (see https://github.com/tensorflow/neural-structured-learning/issues/87#issuecomment-871964966).

Here are a few of our reasons for keeping it in `examples/notebooks` for now:

- we want to let the tutorial mature for some time
- tutorials shown on the TF website directly impose additional maintenance work for our team because these tutorials go through regular testing and underlying TF changes can often break them
- it's harder to remove a tutorial from our website than it is to add a new one

As mentioned in https://github.com/tensorflow/neural-structured-learning/issues/99, I am planning to add a link to the `examples` directory from our TF website. So, that should address some of the visibility concerns.

sayakpaul commented:
> we want to let the tutorial mature for some time

Can you expand a bit more on this? 

If the notebook is going to be hosted on that directory, I'd prefer to add my name (as the author) just after the notebook title. Happy to add acknowledgments too. Here are a few other suggestions:

* A short README [here](https://github.com/tensorflow/neural-structured-learning/tree/master/neural_structured_learning/examples) to clarify the purpose/objectives of the directory so that the readers understand what differentiates the content hosted there from the tutorials. Would also make sense to add one-liners for all the major notebooks/scripts to specify what they are all about. This makes the navigation experience better.
* Mentioning the `examples` directory from the [README](https://github.com/tensorflow/neural-structured-learning/blob/master/README.md) preferably after this section:
   
  ![image](https://user-images.githubusercontent.com/22957388/130712119-44fd18d5-8d19-4484-8d9d-324543fa1824.png)
  I think this is particularly important. Otherwise, there's no way a user could know about it from the repository. 
* Making it clear from this section on [Contribution](https://github.com/tensorflow/neural-structured-learning#contributing-to-nsl). I understand if different tutorial proposals would have different scopes but it still helps to be aware of all these before a proposal.

arjung commented:
Hi Sayak, I was on vacation for the past week and just got back today. Will look at the latest diff within the next couple of days.

arjung reviewed (changes requested):
Hi Sayak,

Since the file was moved, I had a hard time looking at the diff from the previous version and so couldn't figure how to write comments next to the code in your colab directly. I've instead enumerated all my comments below:

- I see the following warning re: the use of the `gaussian_random_matrix` function. Is there anything that can be done to avoid this now?
```
/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function gaussian_random_matrix is deprecated; gaussian_random_matrix is deprecated in 0.22 and will be removed in version 0.24.
  warnings.warn(msg, category=FutureWarning)
```
- Can you disable saving of output in the colab cells?
- Discrepancy in the number of graph edges between what's mentioned in the text and what's in the final graph tsv file. The output of that cell contains 987246 as the value whereas the text mentions 132117
- Missing blank line before the `create_records` function. Convention is 2 blank lines between function definitions
- Do you use unlabeled examples or not in the example? The invocation to `pack_nbrs()` suggests that you don't but in the text before that, you mention that you do. So it's a bit confusing.
 - Make `num_train_examples` and `num_validation_examples` separate bullet points in the `Hyperparameters` section
 - Optional: consider moving the neighbor visualization section soon after building the graph/augmenting the training data and before training the base model.
 - Avoid saying the 'ablation study reveals a different trend' and 'the trade-off reverses'. Just say 'here is ablation study with respect to various values of embedding dimensionality' and include the table. That should be sufficient.
- In the acknowledgement section, I believe you meant to say 'would not have been possible'

These are all minor comments but the colab looks in great shape overall. We should be ready to merge the next version you send.

sayakpaul commented:
@arjung not sure why the notebook commenting was troublesome but I appreciate you being thorough with the comments. 

I have addressed all the comments except for this one:

> Do you use unlabeled examples or not in the example? The invocation to pack_nbrs() suggests that you don't but in the text before that, you mention that you do. So it's a bit confusing.

I don't think I mentioned this. However, if you could point me to any specific line that indicates this, happy to look into that again.

arjung commented:
> > Do you use unlabeled examples or not in the example? The invocation to pack_nbrs() suggests that you don't but in the text before that, you mention that you do. So it's a bit confusing.
> 
> I don't think I mentioned this. However, if you could point me to any specific line that indicates this, happy to look into that again.

Sorry my bad. I misread 'undirected' as 'unlabeled' in the text before the call to `pack_nbrs()`. Ignore this comment. Everything looks good now!
